File: app/src/main/java/io/sweers/catchup/data/smmry/model/SmmryRequestBuilder.java
package io.sweers.catchup.data.smmry.model;

import io.sweers.catchup.BuildConfig;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.LinkedHashMap;
import java.util.Map;

public final class SmmryRequestBuilder {

  public static SmmryRequestBuilder forUrl(String url) {
    SmmryRequestBuilder builder = new SmmryRequestBuilder();
    try {
      builder.url = URLEncoder.encode(url, "UTF-8");
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException("Invalid url - " + url);
    }
    return builder;
  }

  public static SmmryRequestBuilder forText() {
    return new SmmryRequestBuilder();
  }

  // the webpage to summarize
  private String url;

  // the number of sentences returned, default is 7
  private long sentenceCount = -1;

  // how many of the top keywords to return
  private long keywordCount = -1;

  // Whether or not to include quotations
  private boolean avoidQuote = false;
  private boolean avoidQuoteWasSet = false;

  // summary will contain string [BREAK] between each sentence
  private boolean withBreak = false;
  private boolean withBreakWasSet = false;

  private SmmryRequestBuilder() {
  }

  public SmmryRequestBuilder sentenceCount(long sentenceCount) {
    this.sentenceCount = sentenceCount;
    return this;
  }

  public SmmryRequestBuilder keywordCount(long keywordCount) {
    this.keywordCount = keywordCount;
    return this;
  }

  public SmmryRequestBuilder avoidQuote(boolean avoidQuote) {
    this.avoidQuote = avoidQuote;
    avoidQuoteWasSet = true;
    return this;
  }

  public SmmryRequestBuilder withBreak(boolean withBreak) {
    this.withBreak = withBreak;
    withBreakWasSet = true;
    return this;
  }

  public Map<String, Object> build() {
    Map<String, Object> map = new LinkedHashMap<>(6);
    map.put("SM_API_KEY", BuildConfig.SMMRY_API_KEY);
    if (sentenceCount != -1) {
      map.put("SM_LENGTH", sentenceCount);
    }
    if (keywordCount != -1) {
      map.put("SM_KEYWORD_COUNT", keywordCount);
    }
    if (avoidQuoteWasSet) {
      map.put("SM_QUOTE_AVOID", avoidQuote);
    }
    if (withBreakWasSet) {
      map.put("SM_WITH_BREAK", withBreak);
    }

    // This has to be last!
    if (url != null) {
      map.put("SM_URL", url);
    }
    return map;
  }
}

File: libraries/smmry/src/main/kotlin/io/sweers/catchup/smmry/model/SmmryRequestBuilder.kt
/*
 * Copyright (C) 2019. Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.sweers.catchup.smmry.model

import io.sweers.catchup.smmry.BuildConfig
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.LinkedHashMap

class SmmryRequestBuilder private constructor() {

  // the webpage to summarize
  private var url: String? = null

  // the number of sentences returned, default is 7
  private var sentenceCount: Long = -1

  // how many of the top keywords to return
  private var keywordCount: Long = -1

  // Whether or not to include quotations
  private var avoidQuote = false
  private var avoidQuoteWasSet = false

  // summary will contain string [BREAK] between each sentence
  private var withBreak = false
  private var withBreakWasSet = false

  fun sentenceCount(sentenceCount: Long): SmmryRequestBuilder {
    this.sentenceCount = sentenceCount
    return this
  }

  fun keywordCount(keywordCount: Long): SmmryRequestBuilder {
    this.keywordCount = keywordCount
    return this
  }

  fun avoidQuote(avoidQuote: Boolean): SmmryRequestBuilder {
    this.avoidQuote = avoidQuote
    avoidQuoteWasSet = true
    return this
  }

  fun withBreak(withBreak: Boolean): SmmryRequestBuilder {
    this.withBreak = withBreak
    withBreakWasSet = true
    return this
  }

  fun build(): Map<String, Any> {
    val map = LinkedHashMap<String, Any>(6)
    map["SM_API_KEY"] = BuildConfig.SMMRY_API_KEY
    if (sentenceCount != -1L) {
      map["SM_LENGTH"] = sentenceCount
    }
    if (keywordCount != -1L) {
      map["SM_KEYWORD_COUNT"] = keywordCount
    }
    if (avoidQuoteWasSet) {
      map["SM_QUOTE_AVOID"] = avoidQuote
    }
    if (withBreakWasSet) {
      map["SM_WITH_BREAK"] = withBreak
    }

    // This has to be last!
    url?.run {
      map.put("SM_URL", this)
    }
    return map
  }

  companion object {

    fun forUrl(url: String, isAlreadyEncoded: Boolean = false): SmmryRequestBuilder {
      val builder = SmmryRequestBuilder()
      try {
        builder.url = if (isAlreadyEncoded) url else URLEncoder.encode(url, "UTF-8")
      } catch (e: UnsupportedEncodingException) {
        throw RuntimeException("Invalid url - $url")
      }

      return builder
    }

    fun forText(): SmmryRequestBuilder {
      return SmmryRequestBuilder()
    }
  }
}
